torch/utils/_pytree.py

```diff
@@ -282,11 +309,44 @@ def _private_register_pytree_node(
         SUPPORTED_SERIALIZED_TYPES[cls] = serialize_node_def
         SERIALIZED_TYPE_TO_PYTHON_TYPE[serialized_type_name] = cls
 
+        if flatten_with_keys_fn is not None:
+            SUPPORTED_KEY_PATH_NODES[cls] = _KeyPathNodeDef(cls, flatten_with_keys_fn)
+
+
+@dataclasses.dataclass(frozen=True)
+class SequenceKey:
+    idx: int
+
+    def __str__(self) -> str:
+        return f"[{self.idx!r}]"
+
+
+@dataclasses.dataclass(frozen=True)
+class DictKey:
+    key: Hashable
+
+    def __str__(self) -> str:
+        return f"[{self.key!r}]"
+
+
+@dataclasses.dataclass(frozen=True)
+class GetAttrKey:
```

Comment:
If this (and the above) are KeyEntry, we should make that typing explicit.

Reply:
I changed KeyEntry to be a hashable protocol, so we're using structural subtyping for this now